kinto/core/storage/postgresql/__init__.py

```diff
@@ -533,9 +533,19 @@ def purge_deleted(self, collection_id, parent_id, before=None,
             placeholders['before'] = before
 
         with self.client.connect() as conn:
-            result = conn.execute(query.format_map(safeholders), placeholders)
+            result = conn.execute(delete_tombstones.format_map(safeholders), placeholders)
+            deleted = result.rowcount
+
+            # If purging everything from a parent_id, then clear timestamps.
+            if collection_id is None and before is None:
+                delete_timestamps = """
+                DELETE
+                FROM timestamps
+                WHERE {parent_id_filter}
+                """
+                conn.execute(delete_timestamps.format_map(safeholders), placeholders)
```

Comment:
What I was wondering is:

1. T1 deletes a collection "record".
2. T2 inserts a collection "record" of the same name and commits. (Because of READ COMMITTED, it doesn't see the collection in T1 anyhow.)
3. T3 looks for the records that are children of that collection, causing a `timestamps` record to be written.
4. T1 then does `purge_deleted` and wipes the timestamps created in T3 (which it sees because of `READ COMMITTED`).

But I'm not really sure I understand SQL isolation well enough to know if this is a problem, and even if it is, we may not really be able to do anything about it. By the time we get to `purge_deleted`, we've already executed the delete query for the collection and deleted all of its records..